### DISCUSSION TITLE
Load/unload time varies depending on current load.

### DISCUSSION THREAD

**Discussion by odashevsky** [google/or-tools#3651] [Routing (and legacy CP) questions]
Hello.

I have a  VRP scenario where load/unload time needs to be taken into account when scheduling. Currently I am padding load/unload time to distance matrix to pad on the load/unload time. However, if vehicle picks up or drops off two or more loads at the same point, the load/unload is actually does not take as long. Namely, the passengers can enter or exist the vehicle at the same time, cutting currently projected the load/unload time in half. How would I go around modeling such scenario?
Thank you.

**Comment by jmarca** [comment]
Do something like checking if the lat lon/location are the same or not when
getting the travel time between nodes.  If node is same location,
you can make travel time a true 0.  If nodes have different locations,
then add on a small value representing fixed cost of changing
locations.

You also know probably if the node is a pickup or dropoff, a priori,
so you can check if a is pickup and b is drop off at same location,
then add the cost, but if a is dropoff and b is dropoff at the same
node, then make it a true zero.

I once did this and ran into the issue that the solver would pick the
cheapest dropoff time.  I think there were nodes for different kinds
of passengers, for example, wheelchair, ambulatory, gurney, etc.  I
had to force it so that gurney came first, then wheelchair, then
ambulatory, because the solver would cheat and make the
gurney/wheelchair load/unload times the same as ambulatory by putting
the ambulatory pickup and dropoff nodes first.

Regards,

James

On Wed, Jan 25, 2023 at 01:57:18PM -0800, Oleg Dashevsky wrote:
> Hello.
>
> I have a  VRP scenario where load/unload time needs to be taken into account when scheduling. Currently I am padding load/unload time to distance matrix to pad on the load/unload time. However, if vehicle picks up or drops off two or more loads at the same point, the load/unload is actually does not take as long. Namely, the passengers can enter or exist the vehicle at the same time, cutting currently projected the load/unload time in half. How would I go around modeling such scenario?
> Thank you.
>
> --
> Reply to this email directly or view it on GitHub:
> https://github.com/google/or-tools/discussions/3651
> You are receiving this because you are subscribed to this thread.
>
> Message ID: ***@***.***>

--

James E. Marca
Activimetrics LLC

**Reply by odashevsky** [reply]
Thanks for the reply. Funny enough, this is exactly the problem I am solving: non-emergency medical transportation of wheelchair and ambulatory passengers. Luckily, I have a liberty to schedule wheelchair passengers separately from ambulatory clients as they use completely different set of vehicles. So for wheelchair it does not matter if multiple clients are being dropped off; wheelchairs require 10 minutes each to unload. Ambulatory, on the other hand, can exit the vehicle at the same time if they're going to the same location. Currently I am adding load/unload time to the "open window" at location based on appointment time to make the window wide enough to allow for load/unload, but sounds like it would be more correct to add load/unload time to trip time instead and manipulate it in time matrix based on is it load or unload as you suggested. You agree?

On somewhat related note: I have to limit of how long client can ride on any given trip. Hence the question here:
[https://github.com/google/or-tools/discussions/3278]
Did you have a problem like that? Did you have to use PARALLEL_CHEAPEST_INSERTION instead of PATH_CHEAPEST_ARC to have the condition
`solver.Add(time_dimension.CumulVar(dropOff) <= time_dimension.CumulVar(pickup) + time_transit[pickup][dropoff] + 30min)`
work and not hang the solver? Thanks.

**Comment by jmarca** [comment]
On the initial solution heuristic, I generally try them all and use
what works.  It is just a first solution and so as long as it gets me
started I don't really care.  That said, the routing solver seems to
have a bug buried deep--the search is not very broad...it gets
stuck pretty easily, and picking one initial heuristic can lead to
better final solution than another.

I much prefer the CPSat solver's steady march to optimality over the
routing solver's endless repeated looping with no improvements.

Anyway, for pickup and delivery problems, I always start first with
parallel cheapest insertion, then try path cheapest arc.  If one of
those two cannot get the solver going, then I probably have to rethink
the formulation.

For the question about limiting trip length, that is pretty easy to
solve.  Just set a condition on the delta dimension (time, space,
whatever) between the dropoff node and the pickup node.  cumul(b)-cumul(a)<X.

In my case, I used the straight line travel time (a to b with no
stops) and did some hacks but as I recall I took the minimum of +45
minutes or 2x the direct travel time.  A 5 minute trip is then at most
10, not 50.  I think there was a regulatory constraint that the trip
could not be more than 45 minutes over the direct trip.

If you need more detailed help, I am a consultant and am available for
work, so feel free to reach out directly.

James

On Thu, Jan 26, 2023 at 09:24:50AM -0800, Oleg Dashevsky wrote:
> Thanks for the reply. Funny enough, this is exactly the problem I am solving: non-emergency medical transportation of wheelchair and ambulatory passengers. Luckily, I have a liberty to schedule wheelchair passengers separately from ambulatory clients as they use completely different set of vehicles. So for wheelchair it does not matter if multiple clients are being dropped off; wheelchairs require 10 minutes each to unload. Ambulatory, on the other hand, can exit the vehicle at the same time if they're going to the same location. Currently I am adding load/unload time to the "open window" at location based on appointment time to make the window wide enough to allow for load/unload, but sounds like it would be more correct to add load/unload time to trip time instead and manipulate it in time matrix based on is it load or unload as you suggested. You agree?
>
> On somewhat related note: I have to limit of how long client can ride on any given trip. Hence the question here:
> [https://github.com/google/or-tools/discussions/3278]
> Did you have a problem like that? Did you have to use PARALLEL_CHEAPEST_INSERTION instead of PATH_CHEAPEST_ARC to have the condition
> `solver.Add(time_dimension.CumulVar(dropOff) <= time_dimension.CumulVar(pickup) + time_transit[pickup][dropoff] + 30min)`
> work and not hang the solver? Thanks.
>
>
>
>
> --
> Reply to this email directly or view it on GitHub:
> https://github.com/google/or-tools/discussions/3651#discussioncomment-4789846
> You are receiving this because you commented.
>
> Message ID: ***@***.***>

--

James E. Marca
Activimetrics LLC

**Comment by odashevsky** [comment]
Hi James, blind.line.drawing@gmail.com is in your profile. Should I contact you via the email?

**Comment by jmarca** [comment]
Yes that is me. Same as on the forum list. JamesOn Jan 26, 2023, at 12:07, Oleg Dashevsky ***@***.***> wrote:﻿
Hi James, ***@***.*** is in your profile. Should I contact you via the email?
